File: backend/app/routers/games.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user
from app.core.rate_limit import check_rate_limit
from app.database import get_db
from app.models.user import User
from app.schemas.game import (
    FreeKickClaimOut,
    FreeKickKickOut,
    FreeKickKickRequest,
    FreeKickStartOut,
    FreeKickStartRequest,
    GameLimitsOut,
    HangmanClaimOut,
    HangmanGuessOut,
    HangmanGuessRequest,
    HangmanStartOut,
    MemoryClaimOut,
    MemoryLeaderboardEntry,
    MemoryStartOut,
    MemorySubmitOut,
    MemorySubmitRequest,
    PairsClaimOut,
    PairsFlipOut,
    PairsFlipRequest,
    PairsStartOut,
    PenaltyClaimOut,
    PenaltyForfeitOut,
    PenaltyKickOut,
    PenaltyKickRequest,
    PenaltyStartOut,
    PenaltyStartRequest,
    PenaltyStatsOut,
    SaboteurClaimOut,
    SaboteurRevealOut,
    SaboteurRevealRequest,
    SaboteurStartOut,
    SaboteurStartRequest,
)
from app.schemas.fut_draft import (
    FutDraftCardArenaActionRequest,
    FutDraftClaimOut,
    FutDraftCoinFlipRequest,
    FutDraftConfigOut,
    FutDraftFormationRequest,
    FutDraftLeaderboardEntry,
    FutDraftOpenSlotRequest,
    FutDraftPenaltyKickRequest,
    FutDraftPickRequest,
    FutDraftRoundOut,
    FutDraftStartOut,
    FutDraftStateOut,
    FutDraftTacticoChoiceRequest,
)
from app.services import (
    free_kick_service,
    fut_draft_service,
    game_limits_service,
    hangman_service,
    memory_game_service,
    pairs_service,
    penalty_service,
    saboteur_service,
)
from app.services.game_config_service import get_config

router = APIRouter(prefix="/games", tags=["games"])


@router.get("/limits", response_model=GameLimitsOut)
async def get_game_limits(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    config = await get_config(db)
    return game_limits_service.get_remaining_plays(user, config)


@router.post("/memory/start", response_model=MemoryStartOut)
async def memory_start(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    check_rate_limit(f"memory_start:{user.id}", max_calls=20, window_seconds=60)
    return await memory_game_service.start_session(db, user)


@router.post("/memory/{session_id}/submit", response_model=MemorySubmitOut)
async def memory_submit(
    session_id: int,
    payload: MemorySubmitRequest,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await memory_game_service.submit_round(db, user, session_id, payload.answer)


@router.post("/memory/{session_id}/end", response_model=MemorySubmitOut)
async def memory_end(session_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await memory_game_service.end_session(db, user, session_id)


@router.post("/memory/{session_id}/claim", response_model=MemoryClaimOut)
async def memory_claim(session_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await memory_game_service.claim_reward(db, user, session_id)


@router.get("/memory/leaderboard", response_model=list[MemoryLeaderboardEntry])
async def memory_leaderboard(db: AsyncSession = Depends(get_db), _user: User = Depends(get_current_user)):
    return await memory_game_service.leaderboard(db)


# --- Saboteur ---

@router.post("/saboteur/start", response_model=SaboteurStartOut)
async def saboteur_start(payload: SaboteurStartRequest = SaboteurStartRequest(), db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    check_rate_limit(f"saboteur_start:{user.id}", max_calls=20, window_seconds=60)
    return await saboteur_service.start_session(db, user, payload.steward_count)


@router.post("/saboteur/{session_id}/reveal", response_model=SaboteurRevealOut)
async def saboteur_reveal(session_id: int, payload: SaboteurRevealRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await saboteur_service.reveal_cell(db, user, session_id, payload.cell_index)


@router.post("/saboteur/{session_id}/end", response_model=SaboteurRevealOut)
async def saboteur_end(session_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await saboteur_service.end_session(db, user, session_id)


@router.post("/saboteur/{session_id}/claim", response_model=SaboteurClaimOut)
async def saboteur_claim(session_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await saboteur_service.claim_reward(db, user, session_id)


# --- Penalty ---

@router.get("/penalty/stats", response_model=PenaltyStatsOut)
async def penalty_stats(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await penalty_service.get_stats(db, user)


@router.post("/penalty/start", response_model=PenaltyStartOut)
async def penalty_start(payload: PenaltyStartRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    check_rate_limit(f"penalty_start:{user.id}", max_calls=20, window_seconds=60)
    return await penalty_service.start_session(db, user, payload.user_card_id)


@router.post("/penalty/{session_id}/kick", response_model=PenaltyKickOut)
async def penalty_kick(session_id: int, payload: PenaltyKickRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await penalty_service.resolve_kick(db, user, session_id, payload.direction)


@router.post("/penalty/{session_id}/claim", response_model=PenaltyClaimOut)
async def penalty_claim(session_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await penalty_service.claim_reward(db, user, session_id)


@router.post("/penalty/{session_id}/forfeit", response_model=PenaltyForfeitOut)
async def penalty_forfeit(session_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await penalty_service.forfeit_session(db, user, session_id)


# --- Free Kick ---

@router.post("/free-kick/start", response_model=FreeKickStartOut)
async def free_kick_start(payload: FreeKickStartRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    check_rate_limit(f"free_kick_start:{user.id}", max_calls=20, window_seconds=60)
    return await free_kick_service.start_session(db, user, payload.user_card_id)


@router.post("/free-kick/{session_id}/kick", response_model=FreeKickKickOut)
async def free_kick_kick(session_id: int, payload: FreeKickKickRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await free_kick_service.resolve_kick(db, user, session_id, payload.elapsed_ms)


@router.post("/free-kick/{session_id}/claim", response_model=FreeKickClaimOut)
async def free_kick_claim(session_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await free_kick_service.claim_reward(db, user, session_id)


# --- Football Hangman ---

@router.post("/hangman/start", response_model=HangmanStartOut)
async def hangman_start(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    check_rate_limit(f"hangman_start:{user.id}", max_calls=20, window_seconds=60)
    return await hangman_service.start_session(db, user)


@router.post("/hangman/{session_id}/guess", response_model=HangmanGuessOut)
async def hangman_guess(session_id: int, payload: HangmanGuessRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await hangman_service.guess_letter(db, user, session_id, payload.letter)


@router.post("/hangman/{session_id}/claim", response_model=HangmanClaimOut)
async def hangman_claim(session_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await hangman_service.claim_reward(db, user, session_id)


# --- Найди пару (card pairs memory match) ---

@router.post("/pairs/start", response_model=PairsStartOut)
async def pairs_start(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    check_rate_limit(f"pairs_start:{user.id}", max_calls=20, window_seconds=60)
    return await pairs_service.start_session(db, user)


@router.post("/pairs/{session_id}/flip", response_model=PairsFlipOut)
async def pairs_flip(session_id: int, payload: PairsFlipRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await pairs_service.flip_card(db, user, session_id, payload.position)


@router.post("/pairs/{session_id}/claim", response_model=PairsClaimOut)
async def pairs_claim(session_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await pairs_service.claim_reward(db, user, session_id)


# --- FUT Draft ---

@router.get("/fut-draft/config", response_model=FutDraftConfigOut)
async def fut_draft_config(db: AsyncSession = Depends(get_db), _user: User = Depends(get_current_user)):
    return await fut_draft_service.get_public_config(db)


@router.get("/fut-draft/leaderboard", response_model=list[FutDraftLeaderboardEntry])
async def fut_draft_leaderboard(db: AsyncSession = Depends(get_db), _user: User = Depends(get_current_user)):
    return await fut_draft_service.leaderboard(db)


@router.post("/fut-draft/start", response_model=FutDraftStartOut)
async def fut_draft_start(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    check_rate_limit(f"fut_draft_start:{user.id}", max_calls=20, window_seconds=60)
    return await fut_draft_service.start_draft(db, user)


@router.post("/fut-draft/{session_id}/formation", response_model=FutDraftStateOut)
async def fut_draft_choose_formation(
    session_id: int, payload: FutDraftFormationRequest,
    db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user),
):
    return await fut_draft_service.choose_formation(db, user, session_id, payload.formation)


@router.post("/fut-draft/{session_id}/slot", response_model=FutDraftStateOut)
async def fut_draft_open_slot(
    session_id: int, payload: FutDraftOpenSlotRequest,
    db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user),
):
    return await fut_draft_service.open_slot(db, user, session_id, payload.slot_code)


@router.post("/fut-draft/{session_id}/pick", response_model=FutDraftStateOut)
async def fut_draft_pick(
    session_id: int, payload: FutDraftPickRequest,
    db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user),
):
    return await fut_draft_service.submit_pick(db, user, session_id, payload.player_id)


@router.post("/fut-draft/{session_id}/match/start", response_model=FutDraftRoundOut)
async def fut_draft_start_match(session_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await fut_draft_service.start_match(db, user, session_id)


@router.post("/fut-draft/{session_id}/card-arena/action", response_model=FutDraftRoundOut)
async def fut_draft_card_arena_action(
    session_id: int, payload: FutDraftCardArenaActionRequest,
    db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user),
):
    return await fut_draft_service.submit_card_arena_action(db, user, session_id, payload.action)


@router.post("/fut-draft/{session_id}/tactico/phase", response_model=FutDraftRoundOut)
async def fut_draft_tactico_phase(
    session_id: int, payload: FutDraftTacticoChoiceRequest,
    db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user),
):
    return await fut_draft_service.submit_tactico_phase(db, user, session_id, payload.choice)


@router.post("/fut-draft/{session_id}/penalty/kick", response_model=FutDraftRoundOut)
async def fut_draft_penalty_kick(
    session_id: int, payload: FutDraftPenaltyKickRequest,
    db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user),
):
    return await fut_draft_service.submit_penalty_kick(db, user, session_id, payload.direction)


@router.post("/fut-draft/{session_id}/coin-flip", response_model=FutDraftRoundOut)
async def fut_draft_coin_flip(
    session_id: int, payload: FutDraftCoinFlipRequest,
    db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user),
):
    return await fut_draft_service.submit_coin_flip(db, user, session_id, payload.choice)


@router.post("/fut-draft/{session_id}/claim", response_model=FutDraftClaimOut)
async def fut_draft_claim(session_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await fut_draft_service.claim_reward(db, user, session_id)

File: frontend/src/api/games.ts
import { api } from "@/lib/api";
import type {
  FreeKickClaimResult,
  FreeKickKickResult,
  FreeKickStartResult,
  FutDraftClaim,
  FutDraftCoinFlipChoice,
  FutDraftConfig,
  FutDraftLeaderboardEntry,
  FutDraftRound,
  FutDraftStart,
  FutDraftState,
  GameLimits,
  HangmanClaimResult,
  HangmanGuessResult,
  HangmanStartResult,
  MatchActionKind,
  MemoryClaimResult,
  MemoryLeaderboardEntry,
  MemoryStart,
  MemorySubmitResult,
  PairsClaimResult,
  PairsFlipResult,
  PairsStartResult,
  PenaltyClaimResult,
  PenaltyDirection,
  PenaltyForfeitResult,
  PenaltyKickResult,
  PenaltyStartResult,
  PenaltyStats,
  SaboteurClaimResult,
  SaboteurRevealResult,
  SaboteurStartResult,
} from "@/types";

export async function fetchGameLimits(): Promise<GameLimits> {
  const { data } = await api.get<GameLimits>("/games/limits");
  return data;
}

export async function startMemoryGame(): Promise<MemoryStart> {
  const { data } = await api.post<MemoryStart>("/games/memory/start");
  return data;
}

export async function submitMemoryRound(sessionId: number, answer: string[]): Promise<MemorySubmitResult> {
  const { data } = await api.post<MemorySubmitResult>(`/games/memory/${sessionId}/submit`, { answer });
  return data;
}

export async function endMemoryGame(sessionId: number): Promise<MemorySubmitResult> {
  const { data } = await api.post<MemorySubmitResult>(`/games/memory/${sessionId}/end`);
  return data;
}

export async function claimMemoryReward(sessionId: number): Promise<MemoryClaimResult> {
  const { data } = await api.post<MemoryClaimResult>(`/games/memory/${sessionId}/claim`);
  return data;
}

export async function fetchMemoryLeaderboard(): Promise<MemoryLeaderboardEntry[]> {
  const { data } = await api.get<MemoryLeaderboardEntry[]>("/games/memory/leaderboard");
  return data;
}

// --- Saboteur ---

export async function startSaboteur(stewardCount: number): Promise<SaboteurStartResult> {
  const { data } = await api.post<SaboteurStartResult>("/games/saboteur/start", { steward_count: stewardCount });
  return data;
}

export async function revealSaboteurCell(sessionId: number, cellIndex: number): Promise<SaboteurRevealResult> {
  const { data } = await api.post<SaboteurRevealResult>(`/games/saboteur/${sessionId}/reveal`, { cell_index: cellIndex });
  return data;
}

export async function endSaboteur(sessionId: number): Promise<SaboteurRevealResult> {
  const { data } = await api.post<SaboteurRevealResult>(`/games/saboteur/${sessionId}/end`);
  return data;
}

export async function claimSaboteurReward(sessionId: number): Promise<SaboteurClaimResult> {
  const { data } = await api.post<SaboteurClaimResult>(`/games/saboteur/${sessionId}/claim`);
  return data;
}

// --- Penalty ---

export async function startPenalty(userCardId: number): Promise<PenaltyStartResult> {
  const { data } = await api.post<PenaltyStartResult>("/games/penalty/start", { user_card_id: userCardId });
  return data;
}

export async function kickPenalty(sessionId: number, direction: PenaltyDirection): Promise<PenaltyKickResult> {
  const { data } = await api.post<PenaltyKickResult>(`/games/penalty/${sessionId}/kick`, { direction });
  return data;
}

export async function claimPenaltyReward(sessionId: number): Promise<PenaltyClaimResult> {
  const { data } = await api.post<PenaltyClaimResult>(`/games/penalty/${sessionId}/claim`);
  return data;
}

export async function forfeitPenalty(sessionId: number): Promise<PenaltyForfeitResult> {
  const { data } = await api.post<PenaltyForfeitResult>(`/games/penalty/${sessionId}/forfeit`);
  return data;
}

export async function fetchPenaltyStats(): Promise<PenaltyStats> {
  const { data } = await api.get<PenaltyStats>("/games/penalty/stats");
  return data;
}

// --- Free Kick ---

export async function startFreeKick(userCardId: number): Promise<FreeKickStartResult> {
  const { data } = await api.post<FreeKickStartResult>("/games/free-kick/start", { user_card_id: userCardId });
  return data;
}

export async function kickFreeKick(sessionId: number, elapsedMs: number): Promise<FreeKickKickResult> {
  const { data } = await api.post<FreeKickKickResult>(`/games/free-kick/${sessionId}/kick`, { elapsed_ms: elapsedMs });
  return data;
}

export async function claimFreeKickReward(sessionId: number): Promise<FreeKickClaimResult> {
  const { data } = await api.post<FreeKickClaimResult>(`/games/free-kick/${sessionId}/claim`);
  return data;
}

// --- Football Hangman ---

export async function startHangman(): Promise<HangmanStartResult> {
  const { data } = await api.post<HangmanStartResult>("/games/hangman/start");
  return data;
}

export async function guessHangmanLetter(sessionId: number, letter: string): Promise<HangmanGuessResult> {
  const { data } = await api.post<HangmanGuessResult>(`/games/hangman/${sessionId}/guess`, { letter });
  return data;
}

export async function claimHangmanReward(sessionId: number): Promise<HangmanClaimResult> {
  const { data } = await api.post<HangmanClaimResult>(`/games/hangman/${sessionId}/claim`);
  return data;
}

// --- Найди пару ---

export async function startPairs(): Promise<PairsStartResult> {
  const { data } = await api.post<PairsStartResult>("/games/pairs/start");
  return data;
}

export async function flipPairsCard(sessionId: number, position: number): Promise<PairsFlipResult> {
  const { data } = await api.post<PairsFlipResult>(`/games/pairs/${sessionId}/flip`, { position });
  return data;
}

export async function claimPairsReward(sessionId: number): Promise<PairsClaimResult> {
  const { data } = await api.post<PairsClaimResult>(`/games/pairs/${sessionId}/claim`);
  return data;
}

// --- FUT Draft ---

export async function fetchFutDraftConfig(): Promise<FutDraftConfig> {
  const { data } = await api.get<FutDraftConfig>("/games/fut-draft/config");
  return data;
}

export async function startFutDraft(): Promise<FutDraftStart> {
  const { data } = await api.post<FutDraftStart>("/games/fut-draft/start");
  return data;
}

export async function chooseFutDraftFormation(sessionId: number, formation: string): Promise<FutDraftState> {
  const { data } = await api.post<FutDraftState>(`/games/fut-draft/${sessionId}/formation`, { formation });
  return data;
}

export async function openFutDraftSlot(sessionId: number, slotCode: string): Promise<FutDraftState> {
  const { data } = await api.post<FutDraftState>(`/games/fut-draft/${sessionId}/slot`, { slot_code: slotCode });
  return data;
}

export async function submitFutDraftPick(sessionId: number, playerId: number): Promise<FutDraftState> {
  const { data } = await api.post<FutDraftState>(`/games/fut-draft/${sessionId}/pick`, { player_id: playerId });
  return data;
}

export async function startFutDraftMatch(sessionId: number): Promise<FutDraftRound> {
  const { data } = await api.post<FutDraftRound>(`/games/fut-draft/${sessionId}/match/start`);
  return data;
}

export async function submitFutDraftCardArenaAction(sessionId: number, action: MatchActionKind): Promise<FutDraftRound> {
  const { data } = await api.post<FutDraftRound>(`/games/fut-draft/${sessionId}/card-arena/action`, { action });
  return data;
}

export async function submitFutDraftTacticoPhase(sessionId: number, choice: string): Promise<FutDraftRound> {
  const { data } = await api.post<FutDraftRound>(`/games/fut-draft/${sessionId}/tactico/phase`, { choice });
  return data;
}

export async function submitFutDraftPenaltyKick(sessionId: number, direction: string): Promise<FutDraftRound> {
  const { data } = await api.post<FutDraftRound>(`/games/fut-draft/${sessionId}/penalty/kick`, { direction });
  return data;
}

export async function submitFutDraftCoinFlip(sessionId: number, choice: FutDraftCoinFlipChoice): Promise<FutDraftRound> {
  const { data } = await api.post<FutDraftRound>(`/games/fut-draft/${sessionId}/coin-flip`, { choice });
  return data;
}

export async function claimFutDraftReward(sessionId: number): Promise<FutDraftClaim> {
  const { data } = await api.post<FutDraftClaim>(`/games/fut-draft/${sessionId}/claim`);
  return data;
}

export async function fetchFutDraftLeaderboard(): Promise<FutDraftLeaderboardEntry[]> {
  const { data } = await api.get<FutDraftLeaderboardEntry[]>("/games/fut-draft/leaderboard");
  return data;
}
